File: backend/app/services/ticker_interest.py
"""CR200-R001 — ticker interest analytics: what tickers/tickets users research and
trade, and their order fulfilment, aggregated per symbol.

Portable SQLAlchemy, same convention as `admin_analytics.py` (runs on melehost
Postgres and the sqlite unit fixture). Reuses `admin_analytics._real_users_clause()`
so counts exclude the same CR035 room-benchmark synthetics, seed-burst rows,
and CR125/DEF227-229 probe users already excluded from every other admin
metric — a per-ticker breakdown built on an unfiltered user set would just be
those same distortions cut a different way.

No raw ticker-lookup/view telemetry exists (no `research_events` table) — the
interest signals available are Room runs (`room_runs`), trades (`sim_trades`),
resting orders (`sim_resting_orders`), and watchlist adds (`sim_watchlists`).
"""

from __future__ import annotations

from datetime import datetime, timedelta, timezone
from typing import Any

from sqlalchemy import func, select

from app.db import get_session
from app.db.models import RoomRunRow, SimRestingOrderRow, SimTradeRow, SimWatchlistRow, User
from app.services.admin_analytics import _real_users_clause
from app.services.sim_engine import OPEN_RESTING_STATES, TERMINAL_RESTING_STATES

# CR200-R001 audit MINOR-1: these were hand-restated string literals with no
# tie to the model's declared vocabulary, so an 8th state (added to
# sim_engine.py's RestingOrderState/TERMINAL_RESTING_STATES) would silently
# fall through every bucket below. Derived from sim_engine.py's own tuples
# instead — the canonical source `SimRestingOrderRow`'s docstring (models.py)
# describes — so this file changes automatically when that one does.
_PENDING_STATES = OPEN_RESTING_STATES
_TERMINAL_NOT_FILLED_STATES = tuple(s for s in TERMINAL_RESTING_STATES if s != "filled")


def _utcnow() -> datetime:
    return datetime.now(timezone.utc)


def top_tickers(days: int = 30, limit: int = 25) -> dict[str, Any]:
    """Room runs grouped by ticker, real users only."""
    since = _utcnow() - timedelta(days=days)
    with get_session() as s:
        rows = s.execute(
            select(
                RoomRunRow.ticker,
                func.count().label("runs"),
                func.count(func.distinct(RoomRunRow.user_id)).label("researchers"),
                func.sum(func.coalesce(RoomRunRow.credit_cost, 0)).label("credits"),
            )
            .select_from(RoomRunRow)
            .join(User, User.id == RoomRunRow.user_id)
            .where(_real_users_clause(), RoomRunRow.triggered_at >= since)
            .group_by(RoomRunRow.ticker)
            .order_by(func.count().desc())
            .limit(limit)
        ).all()
        status_rows = s.execute(
            select(RoomRunRow.ticker, RoomRunRow.status, func.count())
            .select_from(RoomRunRow)
            .join(User, User.id == RoomRunRow.user_id)
            .where(_real_users_clause(), RoomRunRow.triggered_at >= since)
            .group_by(RoomRunRow.ticker, RoomRunRow.status)
        ).all()

    by_status: dict[str, dict[str, int]] = {}
    for ticker, status, n in status_rows:
        by_status.setdefault(ticker, {})[status] = n

    return {
        "days": days,
        "tickers": [
            {
                "ticker": ticker,
                "runs": runs,
                "researchers": researchers,
                "credits_spent": int(credits or 0),
                "by_status": by_status.get(ticker, {}),
            }
            for ticker, runs, researchers, credits in rows
        ],
    }


def ticker_orders(days: int = 30, limit: int = 25) -> dict[str, Any]:
    """Orders grouped by ticker: filled vs. pending vs. terminal-not-filled."""
    since = _utcnow() - timedelta(days=days)
    with get_session() as s:
        filled_rows = s.execute(
            select(SimTradeRow.ticker, func.count())
            .select_from(SimTradeRow)
            .join(User, User.id == SimTradeRow.user_id)
            .where(_real_users_clause(), SimTradeRow.opened_at >= since)
            .group_by(SimTradeRow.ticker)
        ).all()
        order_rows = s.execute(
            select(SimRestingOrderRow.ticker, SimRestingOrderRow.state, func.count())
            .select_from(SimRestingOrderRow)
            .join(User, User.id == SimRestingOrderRow.user_id)
            .where(_real_users_clause(), SimRestingOrderRow.placed_at >= since)
            .group_by(SimRestingOrderRow.ticker, SimRestingOrderRow.state)
        ).all()

    def _blank() -> dict[str, int]:
        return {"filled": 0, "pending": 0, "not_filled": 0, "other": 0}

    by_ticker: dict[str, dict[str, int]] = {}
    for ticker, n in filled_rows:
        by_ticker.setdefault(ticker, _blank())
        by_ticker[ticker]["filled"] += n
    for ticker, state, n in order_rows:
        by_ticker.setdefault(ticker, _blank())
        if state == "filled":
            by_ticker[ticker]["filled"] += n
        elif state in _PENDING_STATES:
            by_ticker[ticker]["pending"] += n
        elif state in _TERMINAL_NOT_FILLED_STATES:
            by_ticker[ticker]["not_filled"] += n
        else:
            # CR200-R001 audit MINOR-1: a state outside the documented seven
            # (models.py SimRestingOrderRow) must never vanish silently — an
            # admin reading "12 filled, 3 pending, 1 not filled" has no way
            # to know a row was dropped. Degrade loudly (CR040) instead:
            # count it, don't drop it. test_all_states_are_bucketed pins the
            # vocabulary so an eighth state fails the suite, not this branch.
            by_ticker[ticker]["other"] += n

    ranked = sorted(
        by_ticker.items(),
        key=lambda kv: sum(kv[1].values()),
        reverse=True,
    )[:limit]

    return {
        "days": days,
        "tickers": [{"ticker": t, **counts} for t, counts in ranked],
    }


def watchlist_interest(days: int = 30, limit: int = 25) -> dict[str, Any]:
    """Most-watchlisted tickers, real users only."""
    since = _utcnow() - timedelta(days=days)
    with get_session() as s:
        rows = s.execute(
            select(SimWatchlistRow.ticker, func.count())
            .select_from(SimWatchlistRow)
            .join(User, User.id == SimWatchlistRow.user_id)
            .where(_real_users_clause(), SimWatchlistRow.added_at >= since)
            .group_by(SimWatchlistRow.ticker)
            .order_by(func.count().desc())
            .limit(limit)
        ).all()
    return {
        "days": days,
        "tickers": [{"ticker": ticker, "watchlist_adds": n} for ticker, n in rows],
    }


def ticker_detail(ticker: str, days: int = 90) -> dict[str, Any]:
    """One symbol's full interest history, real users only."""
    since = _utcnow() - timedelta(days=days)
    with get_session() as s:
        runs = s.execute(
            select(RoomRunRow.id, RoomRunRow.user_id, RoomRunRow.status,
                   RoomRunRow.triggered_at, RoomRunRow.credit_cost)
            .select_from(RoomRunRow)
            .join(User, User.id == RoomRunRow.user_id)
            .where(_real_users_clause(), RoomRunRow.ticker == ticker,
                   RoomRunRow.triggered_at >= since)
            .order_by(RoomRunRow.triggered_at.desc())
        ).all()
        trades = s.execute(
            select(SimTradeRow.id, SimTradeRow.user_id, SimTradeRow.side,
                   SimTradeRow.quantity, SimTradeRow.status, SimTradeRow.opened_at)
            .select_from(SimTradeRow)
            .join(User, User.id == SimTradeRow.user_id)
            .where(_real_users_clause(), SimTradeRow.ticker == ticker,
                   SimTradeRow.opened_at >= since)
            .order_by(SimTradeRow.opened_at.desc())
        ).all()
        orders = s.execute(
            select(SimRestingOrderRow.id, SimRestingOrderRow.user_id,
                   SimRestingOrderRow.side, SimRestingOrderRow.state,
                   SimRestingOrderRow.placed_at)
            .select_from(SimRestingOrderRow)
            .join(User, User.id == SimRestingOrderRow.user_id)
            .where(_real_users_clause(), SimRestingOrderRow.ticker == ticker,
                   SimRestingOrderRow.placed_at >= since)
            .order_by(SimRestingOrderRow.placed_at.desc())
        ).all()
        watchlist_adds = s.execute(
            select(func.count())
            .select_from(SimWatchlistRow)
            .join(User, User.id == SimWatchlistRow.user_id)
            .where(_real_users_clause(), SimWatchlistRow.ticker == ticker)
        ).scalar_one()

    return {
        "ticker": ticker,
        "days": days,
        "room_runs": [
            {
                "id": str(rid), "user_id": str(uid), "status": status,
                "triggered_at": ts.isoformat() if ts else None,
                "credit_cost": cost,
            }
            for rid, uid, status, ts, cost in runs
        ],
        "trades": [
            {
                "id": str(tid), "user_id": str(uid), "side": side,
                "quantity": float(qty), "status": status,
                "opened_at": ts.isoformat() if ts else None,
            }
            for tid, uid, side, qty, status, ts in trades
        ],
        "resting_orders": [
            {
                "id": str(oid), "user_id": str(uid), "side": side, "state": state,
                "placed_at": ts.isoformat() if ts else None,
            }
            for oid, uid, side, state, ts in orders
        ],
        "watchlist_adds": watchlist_adds,
    }

File: backend/app/static/admin/ticker_interest.js
/* AMI Trade — Admin console TICKER INTEREST view (CR200-R001).
   Which tickers/tickets users research (room_runs) and trade
   (sim_trades / sim_resting_orders fill state), plus watchlist adds.
   Own tab, separate from STATS (analytics.js) — different grain, per-symbol
   not platform-wide. Real-user counts only (CR035 synthetics / seed
   fixtures / probe users excluded server-side, same as every other tab). */

function showTickerInterest() {
  currentUser = null;
  setTab('ticker-interest');
  document.getElementById('main').innerHTML = `
    <div class="card"><h3><span class="accent">◇</span> MOST RESEARCHED (ROOM RUNS, 30D)</h3><div id="tiTop"></div></div>
    <div class="card"><h3><span class="accent">◇</span> ORDER FULFILMENT BY TICKER (30D)</h3><div id="tiOrders"></div></div>
    <div class="card"><h3><span class="accent">◇</span> WATCHLIST INTEREST (30D)</h3><div id="tiWatchlist"></div></div>
    <div class="card" id="tiDetailCard" style="display:none">
      <h3><span class="accent">◇</span> TICKER DETAIL — <span id="tiDetailTicker"></span></h3>
      <div id="tiDetail"></div>
    </div>
  `;
  loadTickerInterest();
}

function tiTable(rows, cols, onClickTicker) {
  if (!rows.length) return `<div class="empty" style="padding:12px">No data in this window.</div>`;
  const head = cols.map(c => `<th>${esc(c.label)}</th>`).join('');
  const body = rows.map(r => `
    <tr class="tiRow" onclick="${onClickTicker ? `loadTickerDetail('${esc(r.ticker)}')` : ''}" style="${onClickTicker ? 'cursor:pointer' : ''}">
      ${cols.map(c => `<td>${c.render ? c.render(r) : esc(r[c.key] ?? '—')}</td>`).join('')}
    </tr>`).join('');
  return `<table class="tiTable" style="width:100%;border-collapse:collapse">
    <thead><tr>${head}</tr></thead><tbody>${body}</tbody></table>`;
}

async function loadTickerInterest() {
  try {
    const t = await api('GET', '/v1/admin/ticker-interest/top?days=30&limit=25');
    document.getElementById('tiTop').innerHTML = tiTable(t.tickers, [
      { key: 'ticker', label: 'ticker', render: r => `<span class="mono">${esc(r.ticker)}</span>` },
      { key: 'runs', label: 'runs' },
      { key: 'researchers', label: 'researchers' },
      { key: 'credits_spent', label: 'credits' },
      { key: 'by_status', label: 'status', render: r => esc(Object.entries(r.by_status).map(([k, n]) => `${k}:${n}`).join(' ') || '—') },
    ], true);
  } catch (e) {
    document.getElementById('tiTop').innerHTML = `<div class="empty" style="color:var(--red)">${esc(e.message)}</div>`;
  }

  try {
    const o = await api('GET', '/v1/admin/ticker-interest/orders?days=30&limit=25');
    document.getElementById('tiOrders').innerHTML = tiTable(o.tickers, [
      { key: 'ticker', label: 'ticker', render: r => `<span class="mono">${esc(r.ticker)}</span>` },
      { key: 'filled', label: 'filled' },
      { key: 'pending', label: 'pending' },
      { key: 'not_filled', label: 'not filled' },
    ], true);
  } catch (e) {
    document.getElementById('tiOrders').innerHTML = `<div class="empty" style="color:var(--red)">${esc(e.message)}</div>`;
  }

  try {
    const w = await api('GET', '/v1/admin/ticker-interest/watchlist?days=30&limit=25');
    document.getElementById('tiWatchlist').innerHTML = tiTable(w.tickers, [
      { key: 'ticker', label: 'ticker', render: r => `<span class="mono">${esc(r.ticker)}</span>` },
      { key: 'watchlist_adds', label: 'adds' },
    ], true);
  } catch (e) {
    document.getElementById('tiWatchlist').innerHTML = `<div class="empty" style="color:var(--red)">${esc(e.message)}</div>`;
  }
}

async function loadTickerDetail(ticker) {
  const card = document.getElementById('tiDetailCard');
  card.style.display = '';
  document.getElementById('tiDetailTicker').textContent = ticker;
  document.getElementById('tiDetail').innerHTML = `<div class="empty">loading…</div>`;
  card.scrollIntoView({ behavior: 'smooth', block: 'nearest' });
  try {
    const d = await api('GET', `/v1/admin/ticker-interest/detail/${encodeURIComponent(ticker)}?days=90`);
    document.getElementById('tiDetail').innerHTML = `
      <div class="stats">
        ${statTile('room runs', d.room_runs.length)}
        ${statTile('trades', d.trades.length)}
        ${statTile('resting orders', d.resting_orders.length)}
        ${statTile('watchlist adds', d.watchlist_adds)}
      </div>
      <div style="margin-top:10px">
        <div class="k" style="font-size:10px;color:var(--text-low);letter-spacing:0.08em;text-transform:uppercase;margin-bottom:4px">recent room runs</div>
        ${tiTable(d.room_runs.slice(0, 10), [
          { key: 'user_id', label: 'user', render: r => shortId(r.user_id) },
          { key: 'status', label: 'status' },
          { key: 'credit_cost', label: 'credits' },
          { key: 'triggered_at', label: 'when', render: r => fmtDate(r.triggered_at) },
        ])}
      </div>
      <div style="margin-top:10px">
        <div class="k" style="font-size:10px;color:var(--text-low);letter-spacing:0.08em;text-transform:uppercase;margin-bottom:4px">recent orders (trades + resting)</div>
        ${tiTable(
          [...d.trades.map(t => ({ ...t, kind: 'filled', when: t.opened_at })),
           ...d.resting_orders.map(o => ({ ...o, kind: o.state, when: o.placed_at }))]
            .sort((a, b) => new Date(b.when) - new Date(a.when)).slice(0, 10),
          [
            { key: 'user_id', label: 'user', render: r => shortId(r.user_id) },
            { key: 'side', label: 'side' },
            { key: 'kind', label: 'outcome' },
            { key: 'when', label: 'when', render: r => fmtDate(r.when) },
          ],
        )}
      </div>
    `;
  } catch (e) {
    document.getElementById('tiDetail').innerHTML = `<div class="empty" style="color:var(--red)">${esc(e.message)}</div>`;
  }
}

registerView({ id: 'ticker-interest', label: 'TICKERS', show: showTickerInterest });
